cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java

```diff
@@ -135,15 +137,16 @@ public DynamicDatasetContext(TransactionContext context, DatasetFramework datase
       dsArguments.putAll(arguments);
     }
 
+    String datasetKey = String.format("%s.%s", name, dsArguments.hashCode());
```

Comment:
Will do. Though needs to be a public class since both `BasicNotificationContext` and `BasicServiceWorkerContext` need access to it since they define the cache.